File: sdks/ruby/test/test_puid.rb
# frozen_string_literal: true

# Integration suite for the Ruby client — runs against a REAL PUID instance.
#
# Point it at a running server with PUID_ENDPOINT (default http://localhost:8799/api,
# i.e. `npm run dev:e2e` from the repo root). The site origin (for dev-login + the
# dashboard API used to mint keys) is derived by stripping the trailing /api, or set
# PUID_ORIGIN explicitly.
#
# Everything a real endpoint can produce is tested live: id generation, decoding,
# quota, 401, 402 (out of quota), 429 (one per second), and the OAuth2
# client-credentials flow. Two cases a live endpoint never produces are exercised
# differently: a non-JSON error body (a one-shot raw TCP server) and a transport
# failure (a real connection to a closed port). When the server is unreachable the
# live tests skip.

require "minitest/autorun"
require "cgi"
require "json"
require "net/http"
require "securerandom"
require "socket"
require "uri"

require "puid"

ENDPOINT = (ENV["PUID_ENDPOINT"] || "http://localhost:8799/api").chomp("/")
ORIGIN = (ENV["PUID_ORIGIN"] || ENDPOINT.sub(%r{/api\z}, "")).chomp("/")

class PuidTest < Minitest::Test
  def server_up?
    uri = URI("#{ENDPOINT}/openapi.json")
    Net::HTTP.start(uri.host, uri.port, use_ssl: uri.scheme == "https") { |h| h.get(uri.request_uri) }
    true
  rescue StandardError
    false
  end

  def require_server
    skip "PUID not reachable at #{ENDPOINT}; run `npm run dev:e2e` from the repo root" unless server_up?
  end

  def uniq_email(tag)
    "#{tag}-#{SecureRandom.hex(6)}@example.com"
  end

  def http(uri, req)
    Net::HTTP.start(uri.host, uri.port, use_ssl: uri.scheme == "https") { |h| h.request(req) }
  end

  def post_json(url, payload, cookie = nil)
    uri = URI(url)
    req = Net::HTTP::Post.new(uri)
    req["Content-Type"] = "application/json"
    req["Accept"] = "application/json"
    req["Cookie"] = cookie if cookie
    req.body = JSON.generate(payload)
    JSON.parse(http(uri, req).body)
  end

  # dev-login (gated by ALLOW_DEV_LOGIN) returns a 302 with the session cookie; we
  # don't follow it (Net::HTTP doesn't), so we capture Set-Cookie directly.
  def dev_cookie(email)
    uri = URI("#{ORIGIN}/auth/dev-login?email=#{CGI.escape(email)}&next=/dashboard")
    res = http(uri, Net::HTTP::Get.new(uri))
    Array(res.get_fields("set-cookie")).map { |c| c.split(";").first }.join("; ")
  end

  def mint_key(cookie)
    post_json("#{ORIGIN}/dashboard/api/team/keys", { label: "ruby-test" }, cookie)["api_key"]
  end

  def seed_usage(cookie, count)
    uri = URI("#{ORIGIN}/dashboard/api/dev/seed-usage?n=#{count}")
    req = Net::HTTP::Get.new(uri)
    req["Cookie"] = cookie
    http(uri, req)
  end

  def register_client(name)
    res = post_json("#{ENDPOINT}/oauth/register", { client_name: name, redirect_uris: ["https://example.test/cb"] })
    [res["client_id"], res["client_secret"]]
  end

  # One-shot TCP server that answers any request with a 502 + non-JSON body.
  def spawn_502_server
    server = TCPServer.new("127.0.0.1", 0)
    port = server.addr[1]
    Thread.new do
      client = server.accept
      begin
        client.readpartial(2048)
      rescue StandardError
        nil
      end
      body = "<html>nope</html>"
      client.write("HTTP/1.1 502 Bad Gateway\r\nContent-Type: text/html\r\n" \
                   "Content-Length: #{body.bytesize}\r\nConnection: close\r\n\r\n#{body}")
      client.close
      server.close
    end
    "http://127.0.0.1:#{port}"
  end

  # --- real service ---------------------------------------------------------

  def test_ids_unique
    require_server
    key = mint_key(dev_cookie(uniq_email("ids")))
    ids = Puid::Client.new(api_key: key, endpoint: ENDPOINT).ids(10)
    assert_equal 10, ids.size
    assert(ids.all? { |s| s.is_a?(String) && !s.empty? })
    assert_equal 10, ids.uniq.size
  end

  def test_id_single
    require_server
    key = mint_key(dev_cookie(uniq_email("single")))
    refute_empty Puid::Client.new(api_key: key, endpoint: ENDPOINT).id
  end

  def test_ordinal_consecutive
    require_server
    key = mint_key(dev_cookie(uniq_email("ord")))
    client = Puid::Client.new(api_key: key, endpoint: ENDPOINT)
    ids = client.ids(2) # one rate-limited request; ordinal is not rate limited
    a = client.ordinal(ids[0])
    b = client.ordinal(ids[1])
    assert_operator a, :>, 0
    assert_equal 1, b - a
  end

  def test_endpoint_trailing_slash
    require_server
    key = mint_key(dev_cookie(uniq_email("slash")))
    refute_empty Puid::Client.new(api_key: key, endpoint: "#{ENDPOINT}/").id
  end

  def test_quota_does_not_spend
    require_server
    key = mint_key(dev_cookie(uniq_email("quota")))
    client = Puid::Client.new(api_key: key, endpoint: ENDPOINT)
    before = client.quota
    after = client.quota
    refute_empty before["plan"]
    assert_equal before["used"], after["used"]
  end

  def test_bad_key_401
    require_server
    err = assert_raises(Puid::Error) { Puid::Client.new(api_key: "puid_live_nope", endpoint: ENDPOINT).id }
    assert_equal 401, err.status
    assert_equal "unauthorized", err.code
  end

  def test_quota_exceeded_402
    require_server
    cookie = dev_cookie(uniq_email("over-quota"))
    seed_usage(cookie, 1000) # free plan = 1000/day
    err = assert_raises(Puid::Error) { Puid::Client.new(api_key: mint_key(cookie), endpoint: ENDPOINT).id }
    assert_equal 402, err.status
  end

  def test_rate_limit_429
    require_server
    key = mint_key(dev_cookie(uniq_email("rate")))
    client = Puid::Client.new(api_key: key, endpoint: ENDPOINT)
    client.id # first request allowed
    err = assert_raises(Puid::Error) { client.id }
    assert_equal 429, err.status
    assert_equal "rate_limited", err.code
  end

  def test_from_client_credentials
    require_server
    id, secret = register_client("ruby-cc-test")
    client = Puid::Client.from_client_credentials(client_id: id, client_secret: secret, endpoint: ENDPOINT)
    assert_equal 2, client.ids(2).size
  end

  def test_from_client_credentials_bad
    require_server
    err = assert_raises(Puid::Error) do
      Puid::Client.from_client_credentials(client_id: "nope", client_secret: "wrong", endpoint: ENDPOINT)
    end
    assert_operator err.status, :>=, 400
  end

  # --- client-side validation -----------------------------------------------

  def test_default_endpoint
    assert_equal "https://puid.dev/api", Puid::DEFAULT_ENDPOINT
  end

  def test_requires_exactly_one_credential
    assert_raises(Puid::Error) { Puid::Client.new }
    assert_raises(Puid::Error) { Puid::Client.new(api_key: "k", access_token: "t") }
  end

  def test_ids_validation
    client = Puid::Client.new(api_key: "k", endpoint: ENDPOINT)
    [0, 11].each do |n|
      err = assert_raises(Puid::Error) { client.ids(n) }
      assert_equal "invalid_count", err.code
    end
  end

  def test_ordinal_validation
    client = Puid::Client.new(api_key: "k", endpoint: ENDPOINT)
    err = assert_raises(Puid::Error) { client.ordinal("") }
    assert_equal "invalid_puid", err.code
  end

  def test_from_client_credentials_requires_args
    assert_raises(Puid::Error) do
      Puid::Client.from_client_credentials(client_id: "only", client_secret: "", endpoint: ENDPOINT)
    end
  end

  # --- cases a live endpoint can't produce -----------------------------------

  def test_non_json_error_body
    base = spawn_502_server
    err = assert_raises(Puid::Error) { Puid::Client.new(api_key: "k", endpoint: base).id }
    assert_equal 502, err.status
    assert_nil err.code
    assert_includes err.message, "HTTP 502"
  end

  def test_network_error
    err = assert_raises(Puid::Error) { Puid::Client.new(api_key: "k", endpoint: "http://127.0.0.1:1/api").id }
    assert_equal "network_error", err.code
  end
end

File: sdks/python/tests/test_puid.py
"""Integration suite for puid-client — runs against a REAL PUID instance.

Point it at a running server with PUID_ENDPOINT (default http://localhost:8799/api,
i.e. ``npm run dev:e2e`` from the repo root). The site origin (for dev-login + the
dashboard API used to mint keys) is derived by stripping the trailing /api, or set
PUID_ORIGIN explicitly.

Everything a real endpoint can produce is tested against the real endpoint: id
generation, decoding, quota, 401, 402 (out of quota), 429 (one per second), and the
OAuth2 client-credentials flow. Only two cases are exercised differently because a
live endpoint never produces them: a non-JSON error body (a fake opener) and a
transport failure (a real connection to a closed port).
"""

import http.client
import http.cookiejar
import io
import json
import os
import secrets
import urllib.error
import urllib.parse
import urllib.request

import pytest

from puid import DEFAULT_ENDPOINT, Puid, PuidError

ENDPOINT = os.environ.get("PUID_ENDPOINT", "http://localhost:8799/api").rstrip("/")
ORIGIN = os.environ.get("PUID_ORIGIN", ENDPOINT.removesuffix("/api")).rstrip("/")


def uniq_email(tag: str) -> str:
    return f"{tag}-{secrets.token_hex(6)}@example.com"


class DevSession:
    """Cookie-jar session against the dev server. dev-login (gated by ALLOW_DEV_LOGIN)
    stands in for a completed Google sign-in, giving a session cookie usable against
    the dashboard API to mint keys and seed usage."""

    def __init__(self, email: str):
        jar = http.cookiejar.CookieJar()
        self.opener = urllib.request.build_opener(urllib.request.HTTPCookieProcessor(jar))
        url = f"{ORIGIN}/auth/dev-login?email={urllib.parse.quote(email)}&next=/dashboard"
        self.opener.open(url)  # follows the redirect; the cookie jar captures the session
        if not list(jar):
            raise RuntimeError("dev-login set no cookie — is ALLOW_DEV_LOGIN=1 on the server?")

    def _post(self, path: str, payload: dict) -> dict:
        req = urllib.request.Request(
            f"{ORIGIN}{path}",
            data=json.dumps(payload).encode(),
            headers={"Content-Type": "application/json", "Accept": "application/json"},
            method="POST",
        )
        with self.opener.open(req) as r:
            return json.loads(r.read().decode())

    def mint_key(self, label: str = "sdk-test") -> str:
        body = self._post("/dashboard/api/team/keys", {"label": label})
        if not body.get("api_key"):
            raise RuntimeError("mint_key failed")
        return body["api_key"]

    def seed_usage(self, n: int) -> None:
        with self.opener.open(f"{ORIGIN}/dashboard/api/dev/seed-usage?n={n}") as r:
            r.read()


def register_client(name: str) -> dict:
    """Dynamic client registration is unauthenticated, so no session is needed."""
    req = urllib.request.Request(
        f"{ENDPOINT}/oauth/register",
        data=json.dumps(
            {"client_name": name, "redirect_uris": ["https://example.test/cb"]}
        ).encode(),
        headers={"Content-Type": "application/json", "Accept": "application/json"},
        method="POST",
    )
    with urllib.request.urlopen(req) as r:
        body = json.loads(r.read().decode())
    if not body.get("client_id") or not body.get("client_secret"):
        raise RuntimeError("register failed")
    return body


@pytest.fixture(scope="session", autouse=True)
def _require_server():
    try:
        urllib.request.urlopen(f"{ENDPOINT}/openapi.json", timeout=5).read()
    except Exception as exc:  # noqa: BLE001
        pytest.skip(
            f"PUID not reachable at {ENDPOINT} ({exc}). Start `npm run dev:e2e` from the repo root."
        )


# --- real service: generation & decoding ------------------------------------


def test_ids_returns_unique_non_empty():
    key = DevSession(uniq_email("ids")).mint_key()
    ids = Puid(api_key=key, endpoint=ENDPOINT).ids(10)
    assert len(ids) == 10
    assert all(isinstance(i, str) and i for i in ids)
    assert len(set(ids)) == 10


def test_id_single():
    key = DevSession(uniq_email("single")).mint_key()
    assert isinstance(Puid(api_key=key, endpoint=ENDPOINT).id(), str)


def test_ordinal_decodes_consecutive():
    key = DevSession(uniq_email("ord")).mint_key()
    puid = Puid(api_key=key, endpoint=ENDPOINT)
    a, b = puid.ids(2)  # one rate-limited request; ordinal() is not rate limited
    oa, ob = puid.ordinal(a), puid.ordinal(b)
    assert isinstance(oa, int) and oa > 0
    assert ob - oa == 1


def test_endpoint_trailing_slash():
    key = DevSession(uniq_email("slash")).mint_key()
    assert Puid(api_key=key, endpoint=ENDPOINT + "/").id()


# --- real service: quota ----------------------------------------------------


def test_quota_does_not_spend():
    key = DevSession(uniq_email("quota")).mint_key()
    puid = Puid(api_key=key, endpoint=ENDPOINT)
    before = puid.quota()
    after = puid.quota()
    assert isinstance(before["plan"], str)
    assert before["used"] == after["used"]
    if before["limit"] is not None:
        assert before["remaining"] <= before["limit"]


# --- real service: error paths ----------------------------------------------


def test_bad_key_401():
    puid = Puid(api_key="puid_live_definitely_not_real", endpoint=ENDPOINT)
    with pytest.raises(PuidError) as exc:
        puid.id()
    assert exc.value.status == 401
    assert exc.value.code == "unauthorized"


def test_quota_exceeded_402():
    session = DevSession(uniq_email("over-quota"))
    session.seed_usage(1000)  # free plan = 1000/day
    puid = Puid(api_key=session.mint_key(), endpoint=ENDPOINT)
    with pytest.raises(PuidError) as exc:
        puid.id()
    assert exc.value.status == 402


def test_rate_limit_429():
    key = DevSession(uniq_email("rate")).mint_key()
    puid = Puid(api_key=key, endpoint=ENDPOINT)
    puid.id()  # first request allowed
    with pytest.raises(PuidError) as exc:
        puid.id()
    assert exc.value.status == 429
    assert exc.value.code == "rate_limited"


# --- real service: OAuth2 (generate on someone else's behalf) ---------------


def test_from_client_credentials():
    reg = register_client("py-cc-test")
    puid = Puid.from_client_credentials(
        client_id=reg["client_id"], client_secret=reg["client_secret"], endpoint=ENDPOINT
    )
    assert len(puid.ids(2)) == 2


def test_from_client_credentials_bad():
    with pytest.raises(PuidError) as exc:
        Puid.from_client_credentials(client_id="nope", client_secret="wrong", endpoint=ENDPOINT)
    assert exc.value.status is not None and exc.value.status >= 400


# --- client-side validation (no network needed) -----------------------------


def test_default_endpoint():
    assert DEFAULT_ENDPOINT == "https://puid.dev/api"


def test_requires_exactly_one_credential():
    with pytest.raises(PuidError):
        Puid()
    with pytest.raises(PuidError):
        Puid(api_key="k", access_token="t")


def test_ids_validation():
    puid = Puid(api_key="k", endpoint=ENDPOINT)
    for bad in (0, 11, 1.5, -3, True):
        with pytest.raises(PuidError) as exc:
            puid.ids(bad)
        assert exc.value.code == "invalid_count"


def test_ordinal_validation():
    puid = Puid(api_key="k", endpoint=ENDPOINT)
    for bad in ("", None):
        with pytest.raises(PuidError) as exc:
            puid.ordinal(bad)
        assert exc.value.code == "invalid_puid"


def test_from_client_credentials_requires_args():
    with pytest.raises(PuidError):
        Puid.from_client_credentials(client_id="only", client_secret="", endpoint=ENDPOINT)


# --- the two cases a live endpoint can't produce ----------------------------


class _NonJsonErrorOpener:
    """Returns a 502 whose body is not JSON (the real API always returns JSON)."""

    def open(self, req, *args, **kwargs):
        raise urllib.error.HTTPError(
            req.full_url,
            502,
            "Bad Gateway",
            http.client.HTTPMessage(),
            io.BytesIO(b"<html>nope</html>"),
        )


def test_non_json_error_body():
    puid = Puid(api_key="k", endpoint=ENDPOINT, opener=_NonJsonErrorOpener())
    with pytest.raises(PuidError) as exc:
        puid.id()
    assert exc.value.status == 502
    assert exc.value.code is None
    assert "HTTP 502" in str(exc.value)


def test_network_error_closed_port():
    puid = Puid(api_key="k", endpoint="http://127.0.0.1:1/api")
    with pytest.raises(PuidError) as exc:
        puid.id()
    assert exc.value.code == "network_error"
